src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,203 @@
+  { intros x hx,
+    rw mem_roots_map h_ne_zero at hx,
+    specialize hc ((ιEE' γ) - (ιEE' (ιFE c)) * x) (begin
+      have f_root := root_left_of_root_gcd hx,
+      rw [eval₂_comp, eval₂_sub, eval₂_mul,eval₂_C, eval₂_C,eval₂_X, eval₂_map] at f_root,
+      exact (mem_roots_map (minimal_polynomial.ne_zero hα)).mpr f_root,
+    end),
+    specialize hc x (begin
+      rw [mem_roots_map (minimal_polynomial.ne_zero hβ), ←eval₂_map],
+      exact root_right_of_root_gcd hx,
+    end),
+    by_contradiction a,
+    apply hc,
+    apply (div_eq_iff (sub_ne_zero.mpr a)).mpr,
+    simp only [algebra.smul_def, ring_hom.map_add, ring_hom.map_mul, ring_hom.comp_apply],
+    ring },
+  rw ← eq_X_sub_C_of_separable_of_root_eq h_ne_zero h_sep h_root h_splits h_roots,
+  transitivity euclidean_domain.gcd (_ : polynomial E) (_ : polynomial E),
+  { dsimp only [p],
+    convert (gcd_map (algebra_map F⟮γ⟯ E)).symm },
+  { simp [map_comp, map_map, ←is_scalar_tower.algebra_map_eq, h],
+    refl },
+end
+
+section primitive_element_same_universe
+universe u
+
+/-- Primitive element theorem for infinite fields. -/
+theorem primitive_element_inf {F E : Type u} [field F] [field E] [algebra F E]
+  [finite_dimensional F E] (F_sep : is_separable F E) (F_inf : infinite F)
+  (n : ℕ) (hn : findim F E = n) : ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  tactic.unfreeze_local_instances,
+  revert F,
+  apply nat.strong_induction_on n,
+  clear n,
+  rintros n ih F hF hFE F_findim F_sep F_inf rfl,
+  by_cases key : ∃ α : E, findim F F⟮α⟯ > 1,
+  { cases key with α hα,
+    haveI Fα_findim : finite_dimensional F⟮α⟯ E := finite_dimensional.right F F⟮α⟯ E,
+    have Fα_dim_lt_F_dim : findim F⟮α⟯ E < findim F E,
+    { rw ← findim_mul_findim F F⟮α⟯ E,
+      nlinarith [show 0 < findim F⟮α⟯ E, from findim_pos, show 0 < findim F F⟮α⟯, from findim_pos], },
+    have Fα_inf : infinite F⟮α⟯ := infinite.of_injective _ (algebra_map F F⟮α⟯).injective,
+    have Fα_sep : is_separable F⟮α⟯ E := is_separable_tower_top_of_is_separable_tower F F⟮α⟯ E F_sep,
+    obtain ⟨β, hβ⟩ := ih _ Fα_dim_lt_F_dim Fα_sep Fα_inf rfl,
+    obtain ⟨γ, hγ⟩ := primitive_element_two_inf α β F_sep,
+    simp only [←adjoin_simple_adjoin_simple, subalgebra.ext_iff, algebra.mem_top, iff_true, *] at *,
+    exact ⟨γ, λ x, hγ algebra.mem_top⟩, },
+  { push_neg at key,
+    rw ← bot_eq_top_of_findim_adjoin_le_one key,
+    exact ⟨0, by rw adjoin_zero⟩, },
+end
+
+/-- Primitive element theorem in same universe. -/
+theorem primitive_element_aux (F E : Type u) [field F] [field E] [algebra F E]
+  [finite_dimensional F E] (F_sep : is_separable F E) : ∃ α : E, F⟮α⟯ = ⊤ :=
+begin
+  by_cases F_finite : nonempty (fintype F),
+  { exact nonempty.elim F_finite (λ h : fintype F, @primitive_element_of_fintype_bot F _ E _ _ h _), },
+  { exact primitive_element_inf F_sep (not_nonempty_fintype.mp F_finite) (findim F E) rfl, },
+end
+
+end primitive_element_same_universe
+
+/-- Complete primitive element theorem. -/
```

Comment:
This is your main result, and this is what you want others to use. Please consider writing a more detailed docstring. Out of context "complete" doesn't make sense. It's only the rush of victory when you read through this file that explains it :stuck_out_tongue_winking_eye: 
It's good to just quote the informal theorem in the docstring, since this is an important result.

Also, I know that it feels like an abomination, but consider renaming this to `exists_adjoin_simple_eq_top` or something like that. Because primitive elements haven't been formally defined in Lean.
Certainly the docstring should mention "primitive element theorem" so that people can find this result!

Reply:
How about this for the comment: "Primitive element theorem: a finite separable filed extension has a primitive element, i.e.
  there is an `α ∈ E` such that `F⟮α⟯ = (⊤ : subalgebra F E)`."

Reply:
Okay, I changed the name of `primitive_element` :cry: Should we also change the names of the lemmas in that file (which right now are called `primitive_element_two_inf` and so on)?

Reply:
Although consistency is good, I'm not so convinced that `exists_adjoin_simple_eq_top_two_inf` is the most readable or useful name. If you're up for some bikeshedding: how about defining something like `@[reducible] def is_primitive_element (K : Type*) [field K] {L : Type*} [field K] [algebra K L] (x : L) := field.adjoin K {x} = ⊤` at the top of the file, and prefix `exists_` in front of `primitive_element` in this file? So we get names like `exists_primitive_element_two_inf`.

Reply:
I also don't really like `exists_adjoin_simple_eq_top_two_inf` and it could be confusing to have lemmas in the file with `primitive_element` in their title when the main result doesn't (and on the other hand, renaming those lemmas could result in some really ugly names). Also, we tried using your suggested definition of `is_primitive_element` and it worked but made proofs a little longer (a lot of times it was necessary to add an extra line to unfold `is_primitive_element` because it was confusing the rewrite tactic) without much obvious benefit (to me at least). So here's my proposal: we name everything `exists_primitive_element` (and similarly `exists_primitive_element_inf` and so on) but otherwise keep the names, etc, as is. If that doesn't seem like a reasonable idea, let us know and we can definitely change it.

Reply:
Also, Thomas has pointed out that `exists_adjoin_simple_eq_top` leaves out the important hypothesis that the field extension is separable, but anything that has `primitive_element` in the name makes that clear.

Reply:
I think `exists_primitive_element` is a good compromise, if we put an implementation note in the module documentation. (I put a suggestion there.)